Add initial support for StylePropertyMap

cdumez opened the pull request:
#### dcf5f3833e1e5ec71762095a6ce457638e96b8a9
<pre>
Add initial support for StylePropertyMap
<a href="https://bugs.webkit.org/show_bug.cgi?id=246828">https://bugs.webkit.org/show_bug.cgi?id=246828</a>

Reviewed by Youenn Fablet.

Add initial support for StylePropertyMap:
- <a href="https://drafts.css-houdini.org/css-typed-om/#stylepropertymap">https://drafts.css-houdini.org/css-typed-om/#stylepropertymap</a>

This patch introduces the CSSStyleRule.styleMap property, which returns a declared StylePropertyMap:
- <a href="https://drafts.css-houdini.org/css-typed-om/#declared-stylepropertymap-objects">https://drafts.css-houdini.org/css-typed-om/#declared-stylepropertymap-objects</a>

For now, we have decent support for get() / has() / clear() / delete().
set() and append() and not yet supported and throw for now.

* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/declared-styleMap-accepts-inherit-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/idlharness-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/set-var-reference-thcrash-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/declared/append.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/declared/clear-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/declared/declared.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/declared/delete-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/declared/delete-shorthand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/declared/get-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/declared/get-shorthand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/declared/getAll-shorthand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/declared/getAll.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/declared/has.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/declared/iterable.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/declared/set-shorthand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/declared/set.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/inline/append.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/inline/clear-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/inline/delete-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/inline/delete-shorthand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/inline/getAll-shorthand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/inline/getAll.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/inline/has.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/inline/iterable.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/inline/set-shorthand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/inline/set.tentative-expected.txt:
* Source/WebCore/Modules/encryptedmedia/MediaKeyStatusMap.h:
(WebCore::MediaKeyStatusMap::createIterator):
* Source/WebCore/Modules/fetch/FetchHeaders.h:
(WebCore::FetchHeaders::createIterator):
* Source/WebCore/Modules/filesystemaccess/FileSystemDirectoryHandle.cpp:
(WebCore::FileSystemDirectoryHandle::createIterator):
* Source/WebCore/Modules/filesystemaccess/FileSystemDirectoryHandle.h:
* Source/WebCore/Modules/webxr/WebXRHand.h:
(WebCore::WebXRHand::createIterator):
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSDOMAsyncIterator.h:
(WebCore::JSDOMAsyncIteratorBase::JSDOMAsyncIteratorBase):
* Source/WebCore/bindings/js/JSDOMIterator.h:
(WebCore::JSDOMIteratorBase::JSDOMIteratorBase):
(WebCore::iteratorForEach):
* Source/WebCore/css/CSSStyleRule.cpp:
(WebCore::CSSStyleRule::CSSStyleRule):
(WebCore::CSSStyleRule::styleMap):
* Source/WebCore/css/CSSStyleRule.h:
* Source/WebCore/css/CSSStyleRule.idl:
* Source/WebCore/css/FontFaceSet.h:
* Source/WebCore/css/typedom/ComputedStylePropertyMapReadOnly.cpp:
(WebCore::ComputedStylePropertyMapReadOnly::get const):
(WebCore::ComputedStylePropertyMapReadOnly::getAll const):
(WebCore::ComputedStylePropertyMapReadOnly::has const):
(WebCore::ComputedStylePropertyMapReadOnly::entries const):
* Source/WebCore/css/typedom/ComputedStylePropertyMapReadOnly.h:
* Source/WebCore/css/typedom/DeclaredStylePropertyMap.cpp: Added.
(WebCore::DeclaredStylePropertyMap::create):
(WebCore::DeclaredStylePropertyMap::DeclaredStylePropertyMap):
(WebCore::DeclaredStylePropertyMap::size const):
(WebCore::DeclaredStylePropertyMap::entries const):
(WebCore::DeclaredStylePropertyMap::propertyValue const):
(WebCore::DeclaredStylePropertyMap::customPropertyValue const):
(WebCore::DeclaredStylePropertyMap::removeProperty):
(WebCore::DeclaredStylePropertyMap::removeCustomProperty):
(WebCore::DeclaredStylePropertyMap::styleRule const):
(WebCore::DeclaredStylePropertyMap::clear):
* Source/WebCore/css/typedom/DeclaredStylePropertyMap.h: Copied from Source/WebCore/css/typedom/ComputedStylePropertyMapReadOnly.h.
* Source/WebCore/css/typedom/StylePropertyMap.cpp: Added.
(WebCore::StylePropertyMap::get const):
(WebCore::StylePropertyMap::getAll const):
(WebCore::StylePropertyMap::has const):
(WebCore::StylePropertyMap::set):
(WebCore::StylePropertyMap::append):
(WebCore::StylePropertyMap::remove):
* Source/WebCore/css/typedom/StylePropertyMap.h:
(WebCore::StylePropertyMap::clearElement):
* Source/WebCore/css/typedom/StylePropertyMap.idl:
* Source/WebCore/css/typedom/StylePropertyMapReadOnly.cpp:
(WebCore::StylePropertyMapReadOnly::reifyValue):
(WebCore::StylePropertyMapReadOnly::customPropertyValueOrDefault):
(WebCore::StylePropertyMapReadOnly::reifyValueToVector):
(WebCore::StylePropertyMapReadOnly::Iterator::Iterator):
* Source/WebCore/css/typedom/StylePropertyMapReadOnly.h:
(WebCore::StylePropertyMapReadOnly::createIterator):
* Source/WebCore/css/typedom/StylePropertyMapReadOnly.idl:
* Source/WebCore/dom/NodeList.h:
(WebCore::NodeList::createIterator):
* Source/WebCore/dom/StyledElement.cpp:
(WebCore::StyledElement::removeInlineStyleProperty):
(WebCore::StyledElement::removeInlineStyleCustomProperty):
* Source/WebCore/dom/StyledElement.h:
* Source/WebCore/html/CustomPaintImage.cpp:
(WebCore::extractComputedProperty):
* Source/WebCore/html/DOMFormData.h:
(WebCore::DOMFormData::createIterator):
* Source/WebCore/html/URLSearchParams.h:
(WebCore::URLSearchParams::createIterator):

Canonical link: <a href="https://commits.webkit.org/255964@main">https://commits.webkit.org/255964@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cef3793154d1e64e7b683bf77ce5af65072975ab

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/94103 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3294 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/24639 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/103728 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/164064 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3312 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31515 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86411 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/99764 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99762 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/2378 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80522 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/29391 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/84304 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/83987 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72344 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37899 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/17806 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35782 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/19073 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39656 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/41648 "Passed tests") | | 
| [❌ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1937 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41597 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/38307 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/css/typedom/DeclaredStylePropertyMap.cpp:
```diff
@@ -0,0 +1,122 @@
+CSSValue* DeclaredStylePropertyMap::customPropertyValue(const AtomString& propertyName) const
+{
+    if (auto* styleRule = this->styleRule())
+        return styleRule->properties().getCustomPropertyCSSValue(propertyName.string()).get();
```
I prefer the style below where we early return the null case.

cdumez replied:
I opted with the one that was less lines, but OK.

youennf commented on Source/WebCore/css/typedom/StylePropertyMap.h:
```diff
@@ -31,7 +31,25 @@ namespace WebCore {
-    virtual void clearElement() = 0;
+    ExceptionOr<RefPtr<CSSStyleValue>> get(ScriptExecutionContext&, const AtomString& property) const override;
+    ExceptionOr<Vector<RefPtr<CSSStyleValue>>> getAll(ScriptExecutionContext&, const AtomString&) const override;
+    ExceptionOr<bool> has(ScriptExecutionContext&, const AtomString&) const override;
```
Can they be final?

cdumez replied:
Had to make some changes to one of the child classes but yes.

youennf commented on Source/WebCore/css/typedom/StylePropertyMap.cpp:
```diff
@@ -0,0 +1,143 @@
+// https://drafts.css-houdini.org/css-typed-om-1/#dom-stylepropertymapreadonly-get
+ExceptionOr<RefPtr<CSSStyleValue>> StylePropertyMap::get(ScriptExecutionContext& context, const AtomString& property) const
+{
+    auto& document = downcast<Document>(context);
```
It would be somehow nice to not do this downcast but I do not know how we could solve this easily.

cdumez replied:
Yes, the problem is that the base class (StylePropertyMapReadOnly) is not only exposed to Window so I cannot use `[CallWith=CurrentDocument]` in its IDL. However, StylePropertyMap is only exposed to Window so this cast is safe.

cdumez committed:
Add initial support for StylePropertyMap
https://bugs.webkit.org/show_bug.cgi?id=246828

Reviewed by Youenn Fablet.

Add initial support for StylePropertyMap:
- https://drafts.css-houdini.org/css-typed-om/#stylepropertymap

This patch introduces the CSSStyleRule.styleMap property, which returns a declared StylePropertyMap:
- https://drafts.css-houdini.org/css-typed-om/#declared-stylepropertymap-objects

For now, we have decent support for get() / has() / clear() / delete().
set() and append() and not yet supported and throw for now.

* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/declared-styleMap-accepts-inherit-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/idlharness-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/set-var-reference-thcrash-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/declared/append.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/declared/clear-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/declared/declared.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/declared/delete-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/declared/delete-shorthand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/declared/get-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/declared/get-shorthand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/declared/getAll-shorthand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/declared/getAll.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/declared/has.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/declared/iterable.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/declared/set-shorthand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/declared/set.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/inline/append.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/inline/clear-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/inline/delete-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/inline/delete-shorthand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/inline/getAll-shorthand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/inline/getAll.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/inline/has.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/inline/iterable.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/inline/set-shorthand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/inline/set.tentative-expected.txt:
* Source/WebCore/Modules/encryptedmedia/MediaKeyStatusMap.h:
(WebCore::MediaKeyStatusMap::createIterator):
* Source/WebCore/Modules/fetch/FetchHeaders.h:
(WebCore::FetchHeaders::createIterator):
* Source/WebCore/Modules/filesystemaccess/FileSystemDirectoryHandle.cpp:
(WebCore::FileSystemDirectoryHandle::createIterator):
* Source/WebCore/Modules/filesystemaccess/FileSystemDirectoryHandle.h:
* Source/WebCore/Modules/webxr/WebXRHand.h:
(WebCore::WebXRHand::createIterator):
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSDOMAsyncIterator.h:
(WebCore::JSDOMAsyncIteratorBase::JSDOMAsyncIteratorBase):
* Source/WebCore/bindings/js/JSDOMIterator.h:
(WebCore::JSDOMIteratorBase::JSDOMIteratorBase):
(WebCore::iteratorForEach):
* Source/WebCore/css/CSSStyleRule.cpp:
(WebCore::CSSStyleRule::CSSStyleRule):
(WebCore::CSSStyleRule::styleMap):
* Source/WebCore/css/CSSStyleRule.h:
* Source/WebCore/css/CSSStyleRule.idl:
* Source/WebCore/css/FontFaceSet.h:
* Source/WebCore/css/typedom/ComputedStylePropertyMapReadOnly.cpp:
(WebCore::ComputedStylePropertyMapReadOnly::get const):
(WebCore::ComputedStylePropertyMapReadOnly::getAll const):
(WebCore::ComputedStylePropertyMapReadOnly::has const):
(WebCore::ComputedStylePropertyMapReadOnly::entries const):
* Source/WebCore/css/typedom/ComputedStylePropertyMapReadOnly.h:
* Source/WebCore/css/typedom/DeclaredStylePropertyMap.cpp: Added.
(WebCore::DeclaredStylePropertyMap::create):
(WebCore::DeclaredStylePropertyMap::DeclaredStylePropertyMap):
(WebCore::DeclaredStylePropertyMap::size const):
(WebCore::DeclaredStylePropertyMap::entries const):
(WebCore::DeclaredStylePropertyMap::propertyValue const):
(WebCore::DeclaredStylePropertyMap::customPropertyValue const):
(WebCore::DeclaredStylePropertyMap::removeProperty):
(WebCore::DeclaredStylePropertyMap::removeCustomProperty):
(WebCore::DeclaredStylePropertyMap::styleRule const):
(WebCore::DeclaredStylePropertyMap::clear):
* Source/WebCore/css/typedom/DeclaredStylePropertyMap.h: Copied from Source/WebCore/css/typedom/ComputedStylePropertyMapReadOnly.h.
* Source/WebCore/css/typedom/StylePropertyMap.cpp: Added.
(WebCore::StylePropertyMap::get const):
(WebCore::StylePropertyMap::getAll const):
(WebCore::StylePropertyMap::has const):
(WebCore::StylePropertyMap::set):
(WebCore::StylePropertyMap::append):
(WebCore::StylePropertyMap::remove):
* Source/WebCore/css/typedom/StylePropertyMap.h:
(WebCore::StylePropertyMap::clearElement):
* Source/WebCore/css/typedom/StylePropertyMap.idl:
* Source/WebCore/css/typedom/StylePropertyMapReadOnly.cpp:
(WebCore::StylePropertyMapReadOnly::reifyValue):
(WebCore::StylePropertyMapReadOnly::customPropertyValueOrDefault):
(WebCore::StylePropertyMapReadOnly::reifyValueToVector):
(WebCore::StylePropertyMapReadOnly::Iterator::Iterator):
* Source/WebCore/css/typedom/StylePropertyMapReadOnly.h:
(WebCore::StylePropertyMapReadOnly::createIterator):
* Source/WebCore/css/typedom/StylePropertyMapReadOnly.idl:
* Source/WebCore/dom/NodeList.h:
(WebCore::NodeList::createIterator):
* Source/WebCore/dom/StyledElement.cpp:
(WebCore::StyledElement::removeInlineStyleProperty):
(WebCore::StyledElement::removeInlineStyleCustomProperty):
* Source/WebCore/dom/StyledElement.h:
* Source/WebCore/html/CustomPaintImage.cpp:
(WebCore::extractComputedProperty):
* Source/WebCore/html/DOMFormData.h:
(WebCore::DOMFormData::createIterator):
* Source/WebCore/html/URLSearchParams.h:
(WebCore::URLSearchParams::createIterator):

Canonical link: https://commits.webkit.org/255964@main
